File: Assets/scenes/warzonescene/script/WaypointController.js
#pragma strict

var waypoints : Vector3[] ; 
var waypointObj : Transform;
private var mapObjects : GameObject[];
private var mapController : MapController;
var currentWaypoint : int = 0;

/* generate waypoint coordinates before anything else wakes up*/



/* FIX THIS SHIT
* Waypoints are overlaying
* also moving the gate after we've spawned it
*/
function Awake () {
	
	
	
	
}


function Start(){
	mapController = GameObject.Find("SceneScripts").GetComponent.<MapController>();
	mapObjects = new GameObject[14];
	
	spawnWaypoint(0, true);
	//mapController.updateObjectList();
	//mapController.updateObjects();
	
	
}

function gateDone(obj : GameObject){
	var ct : int = 0;
	for(var g: GameObject in mapObjects){
		if (g == obj){
			g.GetComponent.<TargettableObject>().highlighted = false;
			break;
		}
		
		ct++;
	}
	
	if(ct < 13){
		mapObjects[ct+1].GetComponent.<DynamicMapObject>().Activate();
		mapObjects[ct+1].GetComponent.<TargettableObject>().highlighted = true;
		var msg : OSCMessage = new OSCMessage("/game/speedrun/beaconDone");
		msg.Append.<int>(ct+1);
		OSCHandler.Instance.SendMessageToAll(msg);
		currentWaypoint = ct+1;
	}
}

/* actually spawn the items in the scene */
function spawnWaypoint(from : int, createObjects : boolean){

	waypoints = new Vector3[14];
	var xc : int = mapController.sectorPos[0];
	var yc : int = mapController.sectorPos[1];
	for (var a : int = from; a < 14; a++){
		if(Random.Range(0,100) < 50){
			xc += Random.Range(-1,1);
		} else {
			yc += Random.Range(-1,1);
		}
		waypoints[a] = Vector3(xc, yc, a + 1);
	}
	if(Random.Range(0,100) < 50){
		xc += Random.Range(-1,1);
	} else {
		yc += Random.Range(-1,1);
	}
	
	var t : int[] = new int[3];
	t[0] = xc;
	t[1] = yc;
	t[2] = 15;
	mapController.spawnGate(t);
	
	//now spawn them
	var ct : int = 0;
	
	var d : DynamicMapObject;
	for(var v  : Vector3  in waypoints){
		if(ct >= from){
			if(createObjects){
				Debug.Log("Creating wp: " + ct);
				var dynObj : Transform = Instantiate(waypointObj, Vector3(10000,10000,10000), Quaternion.identity);
				d = dynObj.GetComponent.<DynamicMapObject>();
				d.sectorCoord[0] = v.x;
				d.sectorCoord[1] = v.y;
				d.sectorCoord[2] = v.z;
				d.originalPosition = Vector3(0,0,400);
				dynObj.GetComponent.<TargettableObject>().objectName = "Waypoint " + (ct + 1);
				mapObjects[ct] = dynObj.gameObject;
				mapController.updateObject(dynObj.gameObject);
			} else {
			
				d = mapObjects[ct].GetComponent.<DynamicMapObject>();;
				d.sectorCoord[0] = v.x;
				d.sectorCoord[1] = v.y;
				d.sectorCoord[2] = v.z;
				d.originalPosition = Vector3(0,0,400);
				mapObjects[ct] = d.gameObject;
				mapController.updateObject(d.gameObject);
			}
			
		}
		ct ++;
		//dynObj.Deactivate();
	}
	
	mapObjects[from].GetComponent.<TargettableObject>().highlighted = true;
	
}

function FixedUpdate () {
	var minDist : float = 100000;
	var closest : int = 0;
	var ct : int = 0;
	var wpPos : Vector3;
	for(var m : GameObject in mapObjects){
		
		wpPos = m.GetComponent.<DynamicMapObject>().getSectorAsVec();
		var shipPos : Vector3 = mapController.getSectorAsVec();
		var dist : float = (shipPos - wpPos).magnitude;
		if(dist < minDist){
			minDist = dist;
			closest = ct;
		}
		ct++;
	}
	//Debug.Log(closest);
	if(closest > currentWaypoint){
		mapObjects[currentWaypoint].GetComponent.<TargettableObject>().highlighted = false;
		currentWaypoint = closest;
		mapObjects[currentWaypoint].GetComponent.<TargettableObject>().highlighted = true;
	}
	
	wpPos = mapObjects[currentWaypoint].GetComponent.<DynamicMapObject>().getSectorAsVec();
	var closestDist : float = (shipPos - wpPos).magnitude;
	if(closestDist > 1.5 && currentWaypoint < 13){
		Debug.Log("Regenerating waypoints dist = " + dist);

		spawnWaypoint(currentWaypoint, false);
	
	}
	
	
	
	
}

File: Assets/scenes/warzonescene/script/WaypointController.cs
using UnityEngine;
using UnityOSC;
using System.Collections.Generic;

/* keep an eye on the waypoints
 * when ship approaches one spawn another in adjacent sector
 */
public class WaypointController:MonoBehaviour{

	int poolMax = 7;
	public Transform waypointObj;
	Transform[] waypointPool ;
	int poolPtr = 0;
	public int currentWaypoint = 0;
	int lastSpawnId = 0;

	MapController mapController;

	List<Transform> pool = new List<Transform> ();



	public void Awake() {
		OSCSystem._instance.RegisterRpcClass (this);
		
	}
	
	
	public void Start(){
		mapController = GameObject.Find("SceneScripts").GetComponent<MapController>();
		//now position them
		waypointPool = new Transform[poolMax];
		//spawn 4 waypoints for the pool
		for (int i = 0; i < poolMax; i++) {
			Transform t = Instantiate<Transform> (waypointObj);
			t.GetComponent<WaypointBehaviour>().setController(this);
			DynamicMapObject dObject = t.GetComponent<DynamicMapObject> ();
			/*dObject.originalPosition = Random.onUnitSphere * 40f;
			dObject.sectorCoord = new int[] {0,0,i};*/


			dObject.OnDeactivate += ObjectDeactivated;
			dObject.OnActivate += ObjectActivated;
			dObject.Deactivate();
			waypointPool[i] = t;
		}


		Vector3 currentPos = Random.onUnitSphere * 40;
		currentPos.z += 1000 + Random.Range (-200,200);
		currentPos.x += Random.Range(-300,300);
		currentPos.y += Random.Range (-300,300);
		Transform tNew = spawnNew (currentPos, 0);
		tNew.GetComponent<TargettableObject>().highlighted = true;
		currentPos.z += 1000 + Random.Range (-200,200);
		currentPos.x += Random.Range(-300,300);
		currentPos.y += Random.Range (-300,300);
		spawnNew (currentPos, 1);
		lastSpawnId = 1;

	}

	[OscCallable]
	public void PlayersGotLost(){
		//re-pool all of the waypoints
		foreach (Transform t in waypointPool) {
			t.GetComponent<DynamicMapObject>().Deactivate();

		}
		Vector3 currentPos = mapController.getShipWorldPosition () + Random.onUnitSphere * 100f;
		currentPos.z += 1000 + Random.Range (-200,200);
		currentPos.x += Random.Range(-300,300);
		currentPos.y += Random.Range (-300,300);
		Transform tNew = spawnNew (currentPos, lastSpawnId);
		tNew.GetComponent<TargettableObject>().highlighted = true;
		lastSpawnId ++;

		currentPos.z += 1000 + Random.Range (-200,200);
		currentPos.x += Random.Range(-300,300);
		currentPos.y += Random.Range (-300,300);
		spawnNew (currentPos, lastSpawnId);
		lastSpawnId ++;

	}


	/* pull a gate out of the pool, move it somewhere nice then activate it
	 * returns: the gate spawned
	 */
	Transform spawnNew(Vector3 currentWorldPos, int id){

		//find a free waypoint, set its world pos to currentworldpos, update with map controller
		Transform newWp = null;
		if (pool.Count > 0) {
			newWp = pool [0];
			pool.RemoveAt (0);
		} else {
		
			Debug.Log ("RUN OUT OF WAYPOINTS IN POOL");
			return null;
		}

		DynamicMapObject dObject = newWp.GetComponent<DynamicMapObject> ();
		dObject.setWorldPosition (currentWorldPos);
		newWp.GetComponent<WaypointBehaviour>().id = id;
		newWp.GetComponent<GeneralTrackableTarget> ().objectName = "Waypoint " + (id+1);

		mapController.updateObject (dObject.gameObject);
		return newWp;
	}

	void ObjectActivated(DynamicMapObject dObj){

	}

	/* called when a live gate is hidden by the map engine. Return it to the pool */
	void ObjectDeactivated(DynamicMapObject dObj){
		//return to pool
		pool.Add (dObj.transform);

	}

	/* called by gates when player passes through them 
     * if we've passed the last in the chain then spawn a new one from the pool
	 */
	public void gateDone(GameObject obj){
		int id = obj.GetComponent<WaypointBehaviour> ().id;
		if (id >= currentWaypoint) {
			currentWaypoint = id;

			obj.GetComponent<TargettableObject>().highlighted = false;
			OSCMessage msg2 = new OSCMessage("/radar/wayPointReached");
			OSCHandler.Instance.SendMessageToAll(msg2);
			if(lastSpawnId <= currentWaypoint){
				//spawn another one further out
				Vector3 lastPos = obj.GetComponent<DynamicMapObject> ().getWorldPosition ();
				lastPos.z += 1000 + Random.Range (-200,200);
				lastPos.x += Random.Range(-300,300);
				lastPos.y += Random.Range (-300,300);
				lastSpawnId ++;
				Transform newTransform = spawnNew (lastPos, lastSpawnId);
				if(newTransform != null){
					newTransform.GetComponent<TargettableObject>().highlighted = true;
				}
			}
		}


	}

	public void FixedUpdate() {

		
		
	}
}
